### PR TITLE
Authorization 04 task4

### PR DESCRIPTION
**Pull request by nisha-inflectionzone**
### Related issue
  - *Issue Id*
  - *Context*
  
### List of changes
  - *change a*
  - *change b*

### Database migration/schema changes (if any)
- *List your migrations here...*

### Self review checklist
  - [ ] Confirm that the changes made by you meet the resolution criteria and nothing else.
  - [ ] Ran all the tests locally to check that you have not introduced any new regression.
  - [ ] If there are any failing tests, you have already fixed them. If not 
  - [ ] Followed coding and styling guidelines.
  - [ ] Made sure the changes are is not overly complex, unreadable and unnecessarily smart. Simplicity wins!
  - [ ] Made sure the comments are not too verbose and not to cryptic.
  - [ ] Made sure the code metrics are not deteriorated.
  - [ ] Made sure you have updated (if necessary)-
    - [ ] Documentation
    - [ ] Translation
    - [ ] Tests/test data
    - [ ] Postman collection
    - [ ] Version
       
### Additional info (if any)
- *Add anything you feel worth mentioning...*

### screensshots (if any)

Original | Updated
:------------------------:|-------------------------------|
** Original screenshot ** | ** Updated screenshot **|

### CODE REVIEW

File: src/api/clinical/careplan/careplan.controller.ts [hunk 1]
```diff
@@ -184,6 +187,18 @@ export class CareplanController extends BaseController {
     };
 
     //#endregion
+    private authorizePerson = async (model: any) => {
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/careplan/careplan.controller.ts:190]: What is the purpose of this?

File: src/api/clinical/complaint/complaint.controller.ts [hunk 2]
```diff
@@ -81,9 +80,9 @@ export class ComplaintController extends BaseController{
     search = async (request: express.Request, response: express.Response): Promise<void> => {
         try {
 
-            const filters = await this._validator.search(request);
-            await this.authorizeUser(request, filters);
-            const searchResults = await this._service.search(filters);
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/complaint/complaint.controller.ts:86]: Why the pattern for the search broken here?

File: src/api/clinical/diagnosis/diagnosis.controller.ts [hunk 3]
```diff
@@ -91,13 +91,12 @@ export class DiagnosisController extends BaseController{
         try {
 
             const domainModel = await this._validator.update(request);
-
             const id: uuid = await this._validator.getParamUuid(request, 'id');
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/diagnosis/diagnosis.controller.ts:94]: This needs to be looked at. There are some problems.

File: src/api/clinical/diagnosis/diagnosis.auth.ts [hunk 4]
```diff
@@ -39,7 +39,7 @@ export class DiagnosisAuth {
     static getById: AuthOptions = {
         ...DefaultAuthOptions,
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/diagnosis/diagnosis.auth.ts:40]: All the ownership should be with Owner. Scope will be tenant.

File: src/api/clinical/diagnosis/diagnosis.controller.ts [hunk 5]
```diff
@@ -91,13 +91,12 @@ export class DiagnosisController extends BaseController{
...
             const existingUser = await this._service.getById(id);
             if (existingUser == null) {
                 throw new ApiError(404, 'Diagnosis record not found.');
             }
-            await this.authorizeUser(request, existingUser.MedicalPractitionerUserId);
+            await this.authorizeOne(request, existingUser.Patient.UserId, existingUser.Patient.TenantId);
             const updatedDiagnosis = await this._service.update(id, domainModel);
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/diagnosis/diagnosis.controller.ts:100]: This needs to be looked at. There are some problems.

File: src/api/clinical/diagnosis/diagnosis.controller.ts [hunk 6]
```diff
@@ -115,12 +114,11 @@ export class DiagnosisController extends BaseController{
         try {
 
             const id: uuid = await this._validator.getParamUuid(request, 'id');
-            await this.authorizeOne(request, null, null);
             const existingUser = await this._service.getById(id);
             if (existingUser == null) {
                 throw new ApiError(404, 'Diagnosis record not found.');
             }
-            
+            await this.authorizeOne(request, existingUser.Patient.UserId, existingUser.Patient.TenantId);
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/diagnosis/diagnosis.controller.ts:121]: This needs to be looked at. There are some problems.

File: src/api/clinical/lab.record/lab.record.controller.ts [hunk 7]
```diff
@@ -137,8 +137,18 @@ export class LabRecordController extends BaseController {
     };
 
     //#endregion
+    private authorizeUser = async (request: express.Request, ownerUserId: uuid) => {
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/lab.record/lab.record.controller.ts:140]: I think authorizeOne does this job internally. No need to add authorize user method.

File: src/api/clinical/medication/medication.consumption/medication.consumption.auth.ts [hunk 8]
```diff
@@ -79,8 +79,8 @@ export class MedicationConsumptionAuth {
     static getSummaryForDay: AuthOptions = {
         ...DefaultAuthOptions,
         Context     : `${this._baseContext}.GetSummaryForDay`,
-        Ownership   : ResourceOwnership.Owner,
-        ActionScope : ActionScope.Tenant,
+        Ownership   : ResourceOwnership.System,
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/medication/medication.consumption/medication.consumption.auth.ts:82]: Medication consumption is for a particular patient. 
Why this was changed?

File: src/api/clinical/medication/medication.consumption/medication.consumption.controller.ts [hunk 9]
```diff
@@ -288,7 +293,6 @@ export class MedicationConsumptionController extends BaseController {
     getSummaryForDay = async (request: express.Request, response: express.Response): Promise<void> => {
         try {
             const model = await MedicationConsumptionValidator.getSummaryForDay(request);
-            await this.authorizeUser(request, model.PatientUserId);
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/medication/medication.consumption/medication.consumption.controller.ts:291]: Why it was removed?

File: src/api/clinical/medication/medication/medication.auth.ts [hunk 10]
```diff
@@ -48,31 +48,31 @@ export class MedicationAuth {
         ...DefaultAuthOptions,
         Context     : `${this._baseContext}.GetAdministrationRoutes`,
         Ownership   : ResourceOwnership.System,
-        ActionScope : ActionScope.Public,
+        ActionScope : ActionScope.Tenant,
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/medication/medication/medication.auth.ts:51]: This is public resource. Should be open for everybody. It basically lists types for giving medicine to the patient. This is common to all for all tenants.

File: src/api/clinical/order/order.controller.ts [hunk 11]
```diff
@@ -125,8 +126,18 @@ export class OrderController extends BaseController {
     };
 
     //#endregion
+    private authorizeUser = async (request: express.Request, ownerUserId: uuid) => {
+        const _userService = Injector.Container.resolve(UserService);
+        const user = await _userService.getById(ownerUserId);
```
Comment (kiran-kharade-inflection) [inline src/api/clinical/order/order.controller.ts:131]: I think authorizeOne is fine here.

File: src/api/wellness/nutrition/food.consumption/food.consumption.controller.ts [hunk 12]
```diff
@@ -142,7 +140,7 @@ export class FoodConsumptionController extends BaseController {
             if (questionnaire.length === 0) {
                 throw new ApiError(400, 'Cannot fetch nutrition questionnaire!');
             }
-
+            await this.authorizeOne(request, null, null);
```
Comment (kiran-kharade-inflection) [inline src/api/wellness/nutrition/food.consumption/food.consumption.controller.ts:143]: ?

File: src/auth/custom/resource.handler.ts [hunk 13]
```diff
@@ -1,65 +1,65 @@
-import express from 'express';
```
Comment (kiran-kharade-inflection) [inline src/auth/custom/resource.handler.ts:1]: Remove the file entirely.

**Review by kiran-kharade-inflection — CHANGES_REQUESTED** [review]
There are some comments I have added. Please look into them.

File: src/api/clinical/diagnosis/diagnosis.auth.ts (hunk 4 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/diagnosis/diagnosis.auth.ts:40]: Yes.

File: src/api/clinical/diagnosis/diagnosis.controller.ts (hunk 3 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/diagnosis/diagnosis.controller.ts:94]: Done.

File: src/api/clinical/diagnosis/diagnosis.controller.ts (hunk 5 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/diagnosis/diagnosis.controller.ts:100]: Diagnosis dto was not returning patient user id. Modifyied the mapper in order to return the patient user id.

File: src/api/clinical/diagnosis/diagnosis.controller.ts (hunk 6 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/diagnosis/diagnosis.controller.ts:121]: Modified the diagnosis mapper

File: src/api/clinical/lab.record/lab.record.controller.ts (hunk 7 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/lab.record/lab.record.controller.ts:140]: Yes, Removed authorizeUser.

File: src/api/clinical/medication/medication.consumption/medication.consumption.auth.ts (hunk 8 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/medication/medication.consumption/medication.consumption.auth.ts:82]: Changed.This endpoint is allowed for all kinds of user. That is the resaon I excluded the consent of resource owner.

File: src/api/clinical/medication/medication/medication.auth.ts (hunk 10 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/medication/medication/medication.auth.ts:51]: Yes, Done chnages.

File: src/api/clinical/order/order.controller.ts (hunk 11 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/order/order.controller.ts:131]: Yes, Removed authorizeUser

File: src/api/wellness/nutrition/food.consumption/food.consumption.controller.ts (hunk 12 above)
Comment (Dattatraya-Londhe) [reply src/api/wellness/nutrition/food.consumption/food.consumption.controller.ts:143]: Done.

File: src/api/clinical/complaint/complaint.controller.ts (hunk 2 above)
Comment (Dattatraya-Londhe) [reply src/api/clinical/complaint/complaint.controller.ts:86]: This endpoint is not like generic search, It's like search by patient user id.

**Pull request merged**